crates/red_knot_python_semantic/src/types.rs

```diff
@@ -567,10 +571,57 @@ impl<'db> Type<'db> {
                 .iter()
                 .all(|&elem_ty| elem_ty.is_subtype_of(db, target)),
 
-            (_, Type::Union(union)) => union
-                .elements(db)
-                .iter()
-                .any(|&elem_ty| self.is_subtype_of(db, elem_ty)),
+            (_, Type::Union(union)) => {
+                // Special handlings
+                // `is_subtype_of(bool, Literal[False] | AlwaysTruthy)`
+                if self.is_subtype_of(db, KnownClass::Bool.to_instance(db)) {
+                    if union
+                        .elements(db)
+                        .iter()
+                        .any(|elem| Type::BooleanLiteral(true).is_subtype_of(db, *elem))
+                        && union
+                            .elements(db)
+                            .iter()
+                            .any(|elem| Type::BooleanLiteral(false).is_subtype_of(db, *elem))
+                    {
+                        return true;
+                    }
+                }
+                // `is_subtype_of(LiteralString, Literal[""] | AlwaysTruthy)`
+                else if self.is_subtype_of(db, Type::LiteralString) {
+                    if union
+                        .elements(db)
+                        .iter()
+                        .any(|elem| Type::string_literal(db, "").is_subtype_of(db, *elem))
+                        && union
+                            .elements(db)
+                            .iter()
+                            .any(|elem| Type::AlwaysTruthy.is_subtype_of(db, *elem))
+                    {
+                        return true;
+                    }
+                }
+                // `is_subtype_of(AlwaysFalsy, Not[LiteralString] | Literal[""])`
+                else if self.is_subtype_of(db, Type::AlwaysFalsy)
+                    || self.is_subtype_of(db, Type::AlwaysTruthy.negate(db))
+                {
+                    if union
+                        .elements(db)
+                        .iter()
+                        .any(|elem| Type::string_literal(db, "").is_subtype_of(db, *elem))
+                        && union
+                            .elements(db)
+                            .iter()
+                            .any(|elem| Type::LiteralString.negate(db).is_subtype_of(db, *elem))
+                    {
+                        return true;
+                    }
+                }
```

Comment:
I think this is the special case that is most concerning from a performance perspective, since many types will be subtypes of `Not[AlwaysTruthy]`, even if they are totally unrelated to literal strings, and we will have to do these extra checks on all of those types.